src/qonnx/transformation/general.py

```diff
@@ -359,3 +362,57 @@ def apply(self, model):
 
         # one iteration is enough
         return (model, False)
+
+
+# Groups inputs by categories, i.e., groups dynamic inputs first, followed by
+# initializers. Keeps order of inputs in each category.
+def group_inputs_by_category(node: NodeProto, model):  # noqa
+    # Select all dynamic inputs, which are those without initializer tensor
+    dynamics = [i for i in node.input if model.get_initializer(i) is None]
+    # Select all input which are initializers, which, by exclusion, are all
+    # those not among the dynamic inputs
+    initializers = [i for i in node.input if i not in dynamics]
+    # Return lists of dynamic anc initializer inputs
+    return dynamics, initializers
+
+
+# Tidy-Up transformation sorting the inputs to all commutative operations to
+# have initializer inputs last
+class SortCommutativeInputsInitializerLast(Transformation):
+    """
+    Sorts inputs of nodes describing commutative operations to have initializer
+    inputs last. This order of inputs is assumed by many other transformations.
+    """
+
+    # Set of supported commutative operations
+    #   TODO: There might be more valid operations
+    SUPPORTED_COMMUTATIVE_OPS = {"Add", "Mul", "And", "Or", "Xor", "Sum"}
+
+    # Applies the transform to a whole model graph
+    def apply(self, model):  # noqa
+        # Get the model graph out of the model wrapper object
+        graph = model.graph
+        # Keep track of whether the graph has been modified
+        graph_modified = False
+        # Iterate all nodes in the graph keeping track of the index
+        for index, node in enumerate(graph.node):
+            # Check whether this node is among the supported
+            if node.op_type in self.SUPPORTED_COMMUTATIVE_OPS:
+                # Group node inputs by category
+                dynamics, initializers = group_inputs_by_category(node, model)
+                # Flatten the grouped input list
+                inputs = [*dynamics, *initializers]
+                # Length of sorted and original input list must match
+                assert len(inputs) == len(node.input)
+                # Reassigned inputs from sorted categories
+                #   Note: ONNX does not allow direct assignment to node.input
```

Comment:
minor remark for future info: although I haven't tried it, I believe it should be possible to assign directly to protobuf containers such as `node.input` by using the `node.input[:] = ...` syntax, similar to this:

https://github.com/fastmachinelearning/qonnx/blob/main/src/qonnx/custom_op/base.py#L127

Reply:
Ok, thank you for pointing this out, I tried this somewhere else and this indeed works. Removed the comment here, but still kept the loop for assigning, as this simultaneously tests whether the graph has been modified (could probably be written less verbose, but I see now reason to change it now).